presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.reservoirsample;
+
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.DoubleStream;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class UnweightedDoubleReservoirSample
+        implements Cloneable
+{
+    private static final int MAX_SAMPLES_LIMIT = 1_000_000;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(WeightedDoubleReservoirSample.class).instanceSize();
+
+    private int count;
+    private double[] samples;
+
+    public UnweightedDoubleReservoirSample(int maxSamples)
+    {
+        checkArgument(maxSamples > 0, "Number of samples must be positive");
+        checkArgument(maxSamples <= MAX_SAMPLES_LIMIT, "Maximum number of reservoir samples is capped");
+
+        this.count = 0;
+        this.samples = new double[maxSamples];
+    }
+
+    private UnweightedDoubleReservoirSample(UnweightedDoubleReservoirSample other)
+    {
+        this.count = other.count;
+        this.samples = new double[other.samples.length];
+        System.arraycopy(other.samples, 0, samples, 0, samples.length);
+    }
+
+    private UnweightedDoubleReservoirSample(int count, double[] samples)
+    {
+        this.count = count;
+        this.samples = samples;
+    }
+
+    public int getMaxSamples()
+    {
+        return samples.length;
+    }
+
+    public void add(double element)
+    {
+        if (++count <= samples.length) {
+            samples[count - 1] = element;
+            return;
+        }
+        int index = ThreadLocalRandom.current().nextInt(0, count);
+        if (index < samples.length) {
+            samples[index] = element;
+        }
+    }
+
+    public void mergeWith(UnweightedDoubleReservoirSample other)
+    {
+        checkArgument(samples.length == other.samples.length, "Max samples must be equal");
+        if (other.count < other.samples.length) {
+            other.stream().forEach(o -> add(o));
+            return;
+        }
+        if (count < samples.length) {
```

Comment:
- add an empty line before this block for readability
- remove `final` from local variables
- changing `other` here makes for an unexpected side effect; it would be better to treat `other` as immutable

```
        if (count < samples.length) {
            long originalCount = count;
            double[] originalSamples = Arrays.copyOf(samples, samples.length);
            
            count = other.count;
            samples = Arrays.copyOf(other.samples, other.samples.length);

            for (int i = 0; i < originalCount; i++) {
                add(originalSamples[i]);
            }
            return;
        }
```

Reply:
Not sure I get it. That's the point of `final UnweightedDoubleReservoirSample target = ((UnweightedDoubleReservoirSample) other.clone());`, no?

Reply:
@atavory I missed the `.clone()` part. Is this the only reason to have `clone()` method?

You don't need the cast and I'd rename target to otherCopy for clarity: `UnweightedDoubleReservoirSample otherCopy = other.clone();`